test/apps/iodemo/io_demo.cc

```diff
@@ -1030,6 +1049,37 @@ class DemoServer : public P2pDemoCommon {
 
 
 class DemoClient : public P2pDemoCommon {
+    class Disconnector : public UcxCallback {
+    public:
+        Disconnector(DemoClient &client, UcxConnection &conn) : _client(client),
+            _conn(&conn) {
+            _conn->disconnect(this);
+        }
+
+        virtual ~Disconnector() {
+            delete _conn;
+        }
+
+        virtual void operator()(ucs_status_t status) {
+            server_info_t &server_info = _client.get_server_info(_conn);
+
+            _client._num_sent -= get_num_uncompleted(server_info);
+
+            // Remove connection pointer
+            _client._server_index_lookup.erase(_conn);
+
+            // Remove active servers entry
+            _client.active_servers_remove(server_info.active_index);
+
+            reset_server_info(server_info);
```

Comment:
can we unite a common class for client/server which will delete conn in operator() and delete self?
seems only this part is really different
for simplicity, the server does not need to define own class. it can --server._curr_state.active_conns before starting disconnect process

Reply:
> can we unite a common class for client/server which will delete conn in operator() and delete self?

then inheritor has to call base::operator() from its implementation and ordering is important since base::operator() calls dtor

> it can --server._curr_state.active_conns before starting disconnect process

ok, I will check this

Reply:
> then inheritor has to call base::operator() from its implementation and ordering is important since base::operator() calls dtor

maybe use another virtual function for the client-specific part